File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-app-detail/src/state/appDetailState.ts
import type { AppDetailPageResult } from "../types/appDetailModels.js";

export interface AppDetailState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: AppDetailPageResult<unknown>;
  readonly error?: string;
}

export function initialAppDetailState(): AppDetailState {
  return { status: "idle" };
}

export function reduceAppDetailState(
  state: AppDetailState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: AppDetailPageResult<unknown> }
    | { type: "failed"; error: string },
): AppDetailState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_app_detail/lib/src/state/app_detail_state.dart
import '../models/app_detail_models.dart';

enum AppDetailStatus { idle, loading, ready, error }

class AppDetailState {
  const AppDetailState({
    this.status = AppDetailStatus.idle,
    this.lastResult,
    this.error,
  });

  final AppDetailStatus status;
  final AppDetailPageResult<Object?>? lastResult;
  final String? error;

  AppDetailState loading() => const AppDetailState(status: AppDetailStatus.loading);

  AppDetailState loaded(AppDetailPageResult<Object?> result) =>
      AppDetailState(status: AppDetailStatus.ready, lastResult: result);

  AppDetailState failed(String message) =>
      AppDetailState(status: AppDetailStatus.error, error: message);
}
